### PR TITLE
Test inside `pytest_runtest_call` hook

### LINKED ISSUE #35: pytest-arraydiff causes PytestReturnNotNoneWarning in Pytest 7.2
The pytest hooks which pytest-arraydiff uses needs to be changed so this plugin can intercept the returned array before Pytest can raise a warning/error. I'll see if I can get it to work the same way as pytest-mpl, and open a PR here.

### PR DESCRIPTION
**Pull request by ConorMacBride**
This changes the plugin's pytest integration to be the same as pytest-mpl, which is able to handle intercepting returned values for further testing.

Fix #35

### CODE REVIEW

**Comment by ConorMacBride** [comment]
@pllim I think the CI workflow is disabled

**Comment by pllim** [comment]
Let me close/reopen.

**Comment by pllim** [comment]
Might have to drop Python 3.7 ?

**Comment by ConorMacBride** [comment]
Could just ignore the warning? Which would you prefer?

**Comment by pllim** [comment]
Ah, true, you could.

**Comment by pllim** [comment]
Thanks. I don't grok the internals here, so I pinged the people who has release rights for review (https://pypi.org/project/pytest-arraydiff).

**Comment by ConorMacBride** [comment]
#35 should only be an issue when pytest-arraydiff is either not installed or not enabled. I think it's also worth updating for the enabled case here because we found that the existing wrapper method didn't test class based tests.

**Comment by pllim** [comment]
> I think it's also worth updating for the enabled case here because we found that the existing wrapper method didn't test class based tests.

Do you plan to do that in this PR? If not, please open a new issue for it. Thanks!

**Comment by ConorMacBride** [comment]
I already fixed that in the first commit. I was just justifying the change which I later realised wasn't necessary to fix the actual issue referenced. 😆 

I've now confirmed that pytest-arraydiff doesn't work with class-based tests on recent pytest versions. You can see in the logs that it only shows the warning for the class test (because it wasn't able to wrap the test *method* for array testing): https://github.com/astropy/pytest-arraydiff/actions/runs/2914788915/jobs/4644034561#step:5:148

I tried modifying the return value of `tests/test_pytest_arraydiff.py::TestClass::test_succeeds_class`, however, it always passed on the `main` branch. Although, with this PR it failed as expected when modified.

**Comment by pllim** [comment]
If you don't hear back from the actual package maintainers, try poke them on Astropy Slack. Thanks!

File: setup.cfg [hunk 1]
```diff
@@ -50,6 +50,9 @@ testpaths = tests
 xfail_strict = true
 markers =
     array_compare: for functions using array comparison
+filterwarnings =
+    error
+    ignore:distutils Version classes are deprecated
```
Comment (bsipocz) [inline setup.cfg:55]: do add a comment for ignore lines like this, e.g. which upstream version should be the minimal when this could be removed.

**Review by bsipocz — COMMENTED** [review]
minor comment only, you need Tom to sign off on the actual code changes

**Comment by astrofrog** [comment]
I will try and review this shortly.

I do wonder if there would be scope to develop a pytest plugin that deals with getting return values that both pytest-mpl and this plugin could depend on to avoid duplicating infrastructure?

**Pull request merged**